Mage/src/main/java/mage/game/GameImpl.java

```diff
@@ -2166,7 +2166,7 @@ public UUID addDelayedTriggeredAbility(DelayedTriggeredAbility delayedAbility, A
         DelayedTriggeredAbility newAbility = delayedAbility.copy();
         newAbility.newId();
         if (source != null) {
-            newAbility.setSourceObjectZoneChangeCounter(getState().getZoneChangeCounter(source.getSourceId()));
```

Comment:
Smells bed, that code used in many places:
![shot_240420_230746](https://github.com/magefree/mage/assets/8344157/86b1bfeb-6bf3-4deb-b3ef-e125e95aa40d)

Reply:
Well it's most likely broken in many places.

When a trigger A creating a delayed trigger B, the delayed trigger B should not have a different zcc if the source of A blinks after the A trigger (and the +1/-1 woraround won't fix it, as each time the permanent blinks, its zcc is +2, so you can create scenarios of +4, +6, +8 with multiple blinks)

Reply:
Well let's me explain one last time:

Fortune is being saddled by its Saddle ability.
Watcher sees the SADDLED_EVENT: has fortune's sourceId and stores its zcc at time of saddle (It is 1 at that time in my unit test), so MOR is `fortuneId|1`

On attack, the Fortune's trigger. It is the same Fortune MOR as when the saddle was activated. The TriggerredAbility for the trigger "Whenever {this} attacks while saddled, [...]" has zcc 1. This is good as it is the same as the MOR for the permanent.

At end of combat, the delayed trigger resolves, and somehow has for zcc 3. This is the problem. And the Delayed Trigger has no info in it to retrieve the Trigger that created it. Trying to use the Permanent's info is wrong because there could be many zcc changes between the 2 triggers.

---

If I start with a Fortune permanent at zcc 3, the "Whenever attack" trigger has zcc 3, and the delayed trigger has zcc 5.

Is there any valid reason for the delayed ability to not have the zcc of the ability that started it?

Reply:
It must be same zcc due rules (see my comments below with related rules list). Example:
> is the same as the controller of that object at the time the replacement effect was applied

The problem with your "fix" -- you can't be ensure that a source ability contains a real zcc (static ability doesn't -- cause it works from start of the game):

![shot_240426_172654](https://github.com/magefree/mage/assets/8344157/1f8ea377-ac40-4a82-ba46-6aa972f0f5ee)

Reply:
Ok, so the proper fix is:

if `source.getSourceObjectZoneChangeCounter()` is a real zcc, that one should be used
else should use the current `getState().getZoneChangeCounter(source.getSourceId())`

Reply:
Yes, that's standard code to sync zcc in all other places.

![shot_240426_173613](https://github.com/magefree/mage/assets/8344157/d41320fd-415c-4ceb-8dc2-8bb1cd4befcc)

Reply:
BTW I saw `Rebound` while debugging here -- maybe #12185 can be fixed with it too.

Reply:
I doubt so, the issue with Rebound is that the Replacement Effect is not applied properly at the time it needs to on moving the split card from STACK to GRAVEYARD.
The part of Rebound that does delayed ability is not the problem there.